dbcon/mysql/ha_mcs.cpp

```diff
@@ -1137,6 +1156,514 @@ void ha_mcs::cond_pop()
+*/
+
+int ha_mcs_cache::repair(THD *thd, HA_CHECK_OPT *check_opt)
+{
+  int error= 0, error2;
+  int something_crashed= is_crashed();
+  DBUG_ENTER("ha_mcs_cache::repair");
+
+  if (cache_handler->is_crashed() || !something_crashed)
+  {
+    /* Delete everything that was not already committed */
+    mysql_file_chsize(cache_handler->file->dfile.file,
+                      cache_handler->file->s->state.state.key_file_length,
+                      0, MYF(MY_WME));
+    mysql_file_chsize(cache_handler->file->s->kfile.file,
+                      cache_handler->file->s->state.state.data_file_length,
+                      0, MYF(MY_WME));
+    error= cache_handler->repair(thd, check_opt);
+  }
+  if (parent::is_crashed() || !something_crashed)
+    if ((error2= parent::repair(thd, check_opt)))
+      error= error2;
+
+  DBUG_RETURN(error);
+}
+
+
+/**
+   Write to cache handler or main table
+*/
+int ha_mcs_cache::write_row(const uchar *buf)
+{
+  if (insert_command)
+    return cache_handler->write_row(buf);
+  return parent::write_row(buf);
+}
+
+
+void ha_mcs_cache::start_bulk_insert(ha_rows rows, uint flags)
+{
+  if (insert_command)
+  {
+    bzero(&cache_handler->copy_info, sizeof(cache_handler->copy_info));
+    return cache_handler->start_bulk_insert(rows, flags);
+  }
+  return parent::start_bulk_insert(rows, flags);
+}
+
+
+int ha_mcs_cache::end_bulk_insert()
+{
+  if (insert_command)
+    return cache_handler->end_bulk_insert();
+  return parent::end_bulk_insert();
+}
+
+/******************************************************************************
+ ha_mcs_cache Plugin code
+******************************************************************************/
+
+static handler *ha_mcs_cache_create_handler(handlerton *hton,
+                                            TABLE_SHARE *table,
+                                            MEM_ROOT *mem_root)
+{
+  return new (mem_root) ha_mcs_cache(hton, table, mem_root);
+}
+
+static plugin_ref plugin_maria;
+
+static int ha_mcs_cache_init(void *p)
+{
+  handlerton *cache_hton;
+  int error;
+
+  cache_hton= (handlerton *) p;
+  cache_hton->create= ha_mcs_cache_create_handler;
+  cache_hton->panic= 0;
+  cache_hton->flags= HTON_NO_PARTITION;
+
+  error= mcs_hton == NULL;                   // Engine must exists!
+
+  if (error)
+  {
+      my_error(HA_ERR_INITIALIZATION, MYF(0),
+               "Could not find storage engine %s", "Columnstore");
+      return error;
+  }
+
+  {
+    LEX_CSTRING name= { STRING_WITH_LEN("Aria") };
+    plugin_maria= ha_resolve_by_name(0, &name, 0);
+    mcs_maria_hton= plugin_hton(plugin_maria);
+    error= mcs_maria_hton == NULL;                   // Engine must exists!
+    if (error)
+      my_error(HA_ERR_INITIALIZATION, MYF(0),
+               "Could not find storage engine %s", name.str);
+  }
+
+  return error;
+}
+
+static int ha_mcs_cache_deinit(void *p)
+{
+  if (plugin_maria)
+  {
+    plugin_unlock(0, plugin_maria);
+    plugin_maria= 0;
```

Comment:
If this is a pointer we better use nullptr instead.

Reply:
Yes this is pointer to a pointer. I have changed it to use a NULL here instead, just to keep the NULL usage consistent with the rest of the source file. Although there are several other places in the code still where we are still using 0.